[RFC] hooks: add support for `.test` files and add some initial tests

mvo5 opened the pull request:
The current hooks do not have a good testing story. To fix this
this PR adds support for "hook/foo.test" files that will be run
at the end of the prime phase during the snap build.

It also adds two initial tests and converts the existing ABI test.
If this looks reasonable we can write tests for the remaining hooks
to ensure we do not regress again.

mvo5 committed:
hooks: add support for `.test` files and add some initial tests

The current hooks do not have a good testing story. To fix this
this PR adds support for "hook/foo.test" files that will be run
at the end of the prime phase during the snap build.

It also adds two initial tests and converts the existing ABI test.
If this looks reasonable we can write tests for the remaining hooks
to ensure we do not regress again.

zyga commented on snapcraft.yaml:
```diff
@@ -26,3 +26,5 @@ parts:
       unset LD_LIBRARY_PATH;
       export PATH="/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/snap/bin";
       snapcraftctl prime
+      # ensure buildin tests are run
```
```suggestion
      # ensure built-in tests are run
```

zyga commented on hooks/003-remove-dbus-machine-id.test:
```diff
@@ -0,0 +1,18 @@
+fi
+
+if [ -s /etc/machine-id ]; then
+    echo "/etc/machine-id must exit but must be empty"
```
```suggestion
    echo "/etc/machine-id exists but must be empty"
```

zyga commented on hooks/003-remove-dbus-machine-id.test:
```diff
@@ -0,0 +1,18 @@
+fi
+
+if [ ! -e /etc/machine-id ]; then
+    echo "/etc/machine-id must exit"
```
```suggestion
    echo "/etc/machine-id must exist"
```

pedronis commented on README.md:
```diff
@@ -10,6 +10,17 @@ To build this snap locally you need snapcraft. The project must be built as real
+are run inside the base image filesystem.
+
+Each hook should have a matching `.test` file that validates that the
+`.chroot` file works as expected. Those `.test` scripts will be run
```
it is not very clear from this description that these tests will be run after building with snapcraft,  if I understand the travis.yml.

Also I wonder whether "sanity" or "check" would be a better name for them and this process.

mvo5 replied:
Thanks! I updated the documentation now. As for `test` vs `check` vs `sanity` I have no super strong opinion. Those are not tests in the traditional sense so a rename does make sense. I feel the upside of using `tests` is that people who write code may look for "tests" when wanting to write tests more readily than looking for `check` or `sanity`.

sil2100 reviewed (commented):
Ok, this looks fine (besides those small typos as highlighted by zyga).

Personally (that's just a preference though) I don't like the idea of having both the hooks and the tests in the same directory. It might get a bit confusing at first glance, since looking at the hooks/ directory you'll see files that are not actual hooks. What I would prefer personally is moving those tests to a separate hook-tests/ or similar directory, still following the same layout for running them as here.

But that's just my 5 cents.

mvo5 committed:
address review feedback

mvo5 committed:
Merge remote-tracking branch 'upstream/master' into more-tests

mvo5 commented:
@sil2100 Thank you! I moved the test files out now, the original "design" was inspired by go (which mixes source and test in a very similar way) but as foundations own this I'm fine accommodating your preferences :)

mvo5 committed:
snapcraft.yaml: fix running hook-tests

Only run hook tests when root otherwise the chroot to the prime
dir will not work.

mvo5 committed:
run hook-tests as user without chroot
